state/factory/workingset.go

```diff
@@ -373,48 +375,49 @@ func (ws *workingSet) pickAndRunActions(
 
 	// initial action iterator
 	blkCtx := protocol.MustGetBlockCtx(ctx)
-	actionIterator := actioniterator.NewActionIterator(actionMap)
-	for {
-		nextAction, ok := actionIterator.Next()
-		if !ok {
-			break
-		}
-		if ctx, err = withActionCtx(ctx, nextAction); err == nil {
-			for _, p := range reg.All() {
-				if validator, ok := p.(protocol.ActionValidator); ok {
-					if err = validator.Validate(ctx, nextAction.Action(), ws); err != nil {
-						break
+	if ap != nil {
+		actionIterator := actioniterator.NewActionIterator(ap.PendingActionMap())
+		for {
+			nextAction, ok := actionIterator.Next()
+			if !ok {
+				break
+			}
+			if ctx, err = withActionCtx(ctx, nextAction); err == nil {
+				for _, p := range reg.All() {
+					if validator, ok := p.(protocol.ActionValidator); ok {
+						if err = validator.Validate(ctx, nextAction.Action(), ws); err != nil {
+							break
+						}
 					}
 				}
 			}
-		}
-		if err != nil {
-			actionIterator.PopAccount()
-			continue
-		}
-		receipt, err := ws.runAction(ctx, nextAction)
-		switch errors.Cause(err) {
-		case nil:
-			// do nothing
-		case action.ErrHitGasLimit:
-			actionIterator.PopAccount()
-			continue
-		default:
-			return nil, nil, errors.Wrapf(err, "Failed to update state changes for selp %x", nextAction.Hash())
-		}
-		// TODO: if the action is not handled by any handler, receipt will be nil.
-		//  in this case, should this action be appended?
-		if receipt != nil {
-			blkCtx.GasLimit -= receipt.GasConsumed
-			ctx = protocol.WithBlockCtx(ctx, blkCtx)
-			receipts = append(receipts, receipt)
-		}
-		executedActions = append(executedActions, nextAction)
+			if err != nil {
+				ap.DeleteAction(nextAction)
```

Comment:
this is the place where an invalid action is deleted during preparing block.

Reply:
would suggest pass a callback func DeleteAction() into factory, if this is the fix needed, to unblock the FB release

then we can do a better refactor of actpool as we planned, passing actpool into factory (and create a wrapper struct minter{f: f, ap: ap}) feels more like a hack for me

Reply:
maybe delete in batch to avoid lock and unlock?

Reply:
> would suggest pass a callback func DeleteAction() into factory, if this is the fix needed, to unblock the FB release
> 
> then we can do a better refactor of actpool as we planned, passing actpool into factory (and create a wrapper struct minter{f: f, ap: ap}) feels more like a hack to me

would prefer pass into struct or interface instead of callback function